### PR TITLE
Output Addresses From Capture for void* and Opaque Struct* Parameters

### PR DESCRIPTION
**Pull request by andrew-lunarg**
Fixes issue #646

Adds three versions of DataPointerDecoderToString() making it clear what
types they service, each one converting the address from the capture
file for the pointer in question rather than using the address of an
object on the stack at decode time as the old template did.

### CODE REVIEW

File: framework/decode/custom_vulkan_ascii_consumer.h [hunk 1]
```diff
@@ -80,18 +80,66 @@ inline std::string EnumPointerDecoderToString(PointerDecoder<VkEnumType>* pObj)
...
+/// `vkCmdSetCheckpointNV`, `IDirectFB*` from
+/// `vkGetPhysicalDeviceDirectFBPresentationSupportEXT`, `struct _screen_window*` from
+/// `kGetPhysicalDeviceScreenPresentationSupportQNX`, `Display*` from
+/// `vkGetPhysicalDeviceXlibPresentationSupportKHR`, and from
+/// `vkAcquireXlibDisplayEXT`, and from
+/// `vkGetRandROutputDisplayEXT`, `struct wl_display*` from
+/// `vkGetPhysicalDeviceWaylandPresentationSupportKHR`, `xcb_connection_t*` from
+/// `vkGetPhysicalDeviceXcbPresentationSupportKHR`, Win32 memory `HANDLE` from
+/// `vkGetMemoryWin32HandlePropertiesKHR`.
+/// The uint64_t is treated as an address in each case.
```
Comment (bradgrantham-lunarg) [inline framework/decode/custom_vulkan_ascii_consumer.h:111]: Really great and helpful comment, thanks

File: framework/decode/custom_vulkan_ascii_consumer.h [hunk 2]
```diff
@@ -80,18 +80,66 @@ inline std::string EnumPointerDecoderToString(PointerDecoder<VkEnumType>* pObj)
...
+{
+    std::string str{ "\"" + util::PtrToString(buffer) + "\"" };
+    return str;
+}
+
+/// Convert pointers to arrays of void pointers to a string of the captured address.
+template <typename T, typename U>
+inline std::string DataPointerDecoderToString(PointerDecoder<T, U>* pDecoder)
+{
+    if (nullptr == pDecoder)
```
Comment (bradgrantham-lunarg) [inline framework/decode/custom_vulkan_ascii_consumer.h:110]: I suggest just a GFXRECON_ASSERT() here and don't bother returning a NULL.

File: framework/decode/custom_vulkan_ascii_consumer.h [hunk 3]
```diff
@@ -80,16 +80,52 @@ inline std::string EnumPointerDecoderToString(PointerDecoder<VkEnumType>* pObj)
...
+        return GFXRECON_TOJSON_NULL;
+    }
+    std::string str{ "\"" + util::PtrToString(pDecoder->GetAddress()) + "\"" };
+    return str;
+}
+
+/// Convert void pointers into a string representation of the address in the capture file.
+inline std::string DataPointerDecoderToString(PointerDecoder<uint8_t>* pDecoder)
+{
+    if (nullptr == pDecoder)
```
Comment (bradgrantham-lunarg) [inline framework/decode/custom_vulkan_ascii_consumer.h:122]: Would this be called for `pData` parameter to `vkGetPipelineCacheData`?  That can legally be `nullptr` so this probably should just print "0x0" or some such for that case.  I wonder if you could just omit this `if` altogether.

**Review by bradgrantham-lunarg — CHANGES_REQUESTED** [review]

File: framework/decode/custom_vulkan_ascii_consumer.h (hunk 3 above)
Comment (andrew-lunarg) [reply framework/decode/custom_vulkan_ascii_consumer.h:122]: In that case the null would be returned from the `PointerDecoder<uint8_t>::GetAddress()`. `pDecoder` here should never be null, This parameter should really be a reference but the code generation delivers it as a pointer so I handle the null case just to be defensive.

File: framework/decode/custom_vulkan_ascii_consumer.h (hunk 2 above)
Comment (andrew-lunarg) [reply framework/decode/custom_vulkan_ascii_consumer.h]: Done

File: framework/decode/custom_vulkan_ascii_consumer.h (hunk 3 above)
Comment (andrew-lunarg) [reply framework/decode/custom_vulkan_ascii_consumer.h]: Done

**Review by bradgrantham-lunarg — APPROVED** [review]

**Pull request merged**